### DISCUSSION TITLE
Auto complete on the first tab if even it's ambiguous

### DISCUSSION THREAD

**Discussion by SangTruongTan** [marlonrichert/zsh-autocomplete#796] [Q&A] [answered]
## Environment

```zsh
% typeset -p1 VENDOR OSTYPE ZSH_PATCHLEVEL _autocomplete__funcfiletrace
typeset VENDOR=apple
export OSTYPE=darwin24.0
typeset ZSH_PATCHLEVEL=zsh-5.9-0-g73d3173
typeset -a _autocomplete__funcfiletrace=(
  /opt/homebrew/share/zsh-autocomplete/zsh-autocomplete.plugin.zsh:4
  /Users/myuser/.zshrc:57
  zsh:0
)
% git -C ~autocomplete log --oneline -n1
824efa8836 (HEAD -> stable, tag: 4.4.4) Merge pull request #18704 from Homebrew/ubuntu-run
```

* Operating system: MacOS Sequoia 15.2 
* Terminal emulator: iTerm

## Steps to reproduce

```zsh
% cd $(mktemp -d)
% mkdir something1
% mkdir something2
# When type "cd" and hit "tab", it will auto-complete entire in the first time even it's ambiguous.
% cd something1/
```

I'm wondering if I can disable that feature or not. It's really annoying to me. I tested it and show the same behavior on Linux. Sometime, it makes faults when I type the path very fast.

**Comment by marlonrichert** [comment]
There's a setting for this and it's [documented in the Readme](https://github.com/marlonrichert/zsh-autocomplete?tab=readme-ov-file#first-insert-the-common-substring).

**Reply by SangTruongTan** [reply]
Could you please specify the right command and detailed guidelines for that?

I've tried to add the following configuration by the souring auto-complete command in my .zshrc. But it's failed.

```
# all Tab widgets
zstyle ':autocomplete:*complete*:*' insert-unambiguous yes

# all history widgets
zstyle ':autocomplete:*history*:*' insert-unambiguous yes

# ^S
zstyle ':autocomplete:menu-search:*' insert-unambiguous yes
```

**Reply by marlonrichert** [reply]
Works fine for me in a clean environment:
```zsh
cd $( mktemp -d )
> .zshrc <<EOF
setopt interactivecomments transientrprompt
PS1='%# '
PS2=
RPS2='%^'
# all Tab widgets
zstyle ':autocomplete:*complete*:*' insert-unambiguous yes

# all history widgets
zstyle ':autocomplete:*history*:*' insert-unambiguous yes

# ^S
zstyle ':autocomplete:menu-search:*' insert-unambiguous yes
source /Users/marlon/Repos/marlonrichert/zsh-autocomplete/zsh-autocomplete.plugin.zsh
EOF
env -i HOME=$PWD PATH=$PATH TERM=$TERM ${TERMINFO:+TERMINFO=$TERMINFO} zsh -d

~TMPDIR/tmp.PC4fD5CnTu/
% touch bar baz
% file # Press Tab here.
% file ba
file
bar  baz
```

**Reply by SangTruongTan** [reply]
> ```shell
> setopt interactivecomments transientrprompt
> PS1='%# '
> PS2=
> RPS2='%^'
> # all Tab widgets
> zstyle ':autocomplete:*complete*:*' insert-unambiguous yes
> 
> # all history widgets
> zstyle ':autocomplete:*history*:*' insert-unambiguous yes
> 
> # ^S
> zstyle ':autocomplete:menu-search:*' insert-unambiguous yes
> ```

How about pressing Tab one more time? In my case, it behaves like below.
% file # Press Tab once
% file ba $ Press Tab once again
% file bar

**Reply by marlonrichert** [reply]
Yes, that's expected behavior. It inserted the unambiguous substring on the first tab. Isn't that what you said you wanted? What else are you expecting?

**Reply by SangTruongTan** [reply]
Is that possible to not complete the ambiguous string and it doesn't mater how many times I press the tab key?

**Reply by marlonrichert** [reply]
What do you mean with "not complete"? What do you want to insert instead? Nothing? Why would you want that?

**Reply by SangTruongTan** [reply]
Okay, I expect that when the string is ambiguous, then the subsequence tabs could behave in one of the options below:
1. The tab will switch between the ambiguous candidates. For instance, I have "bar" and "baz" then the tab will show those two options in a circular.
2. Similar to the default behavior of zsh or bash, hitting tab on the ambiguous string will just show the candidate options possible on the screen.

**Reply by marlonrichert** [reply]
> The tab will switch between the ambiguous candidates. For instance, I have "bar" and "baz" then the tab will show those two options in a circular.

That's documented here: https://github.com/marlonrichert/zsh-autocomplete/blob/main/README.md#make-tab-and-shifttab-cycle-completions-on-the-command-line

> Similar to the default behavior of zsh or bash, hitting tab on the ambiguous string will just show the candidate options possible on the screen.

You're already seeing those, because you are using autocompletion. Perhaps you just shouldn't use Autocomplete?

**Reply by SangTruongTan** [reply, marked as answer]
It looks my desire could be satisfied by this configuration. Thanks!
```
bindkey              '^I' menu-select
bindkey "$terminfo[kcbt]" menu-select
```

**Comment by F1248** [comment]
I'm still expiring this on macOS 15.5 (24F74) after following https://github.com/marlonrichert/zsh-autocomplete/blob/main/README.md#first-insert-the-common-substring

Complete `.zshrc`:
```
zstyle ':autocomplete:*complete*:*' insert-unambiguous yes
zstyle ':autocomplete:*history*:*' insert-unambiguous yes
zstyle ':autocomplete:menu-search:*' insert-unambiguous yes

source /opt/homebrew/share/zsh-autocomplete/zsh-autocomplete.plugin.zsh
```

Steps to reproduce:
1. type `n`
2. press `tab`
Expected result: not completion, as `n` is ambiguous
Actual result: `n` is completed to `nano`

On the other hand, path completion works fine. Pressing tab the first time after typing `ls D` does not complete anything.
